sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/GroupIntoBatches.java

```diff
@@ -398,11 +424,31 @@ public void processElement(
         storedBatchSizeBytes.readLater();
       }
 
-      long num = storedBatchSize.read();
-      if (maxBufferingDuration.isLongerThan(Duration.ZERO) && num == 1) {
-        // This is the first element in batch. Start counting buffering time if a limit was set.
-        bufferingTimer.offset(maxBufferingDuration).setRelative();
+      long num;
+      if (maxBufferingDuration.isLongerThan(Duration.ZERO)) {
+        minBufferedTs.readLater();
+        num = storedBatchSize.read();
+
+        long oldOutputTs =
+            MoreObjects.firstNonNull(
+                minBufferedTs.read(), BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis());
+        minBufferedTs.add(elementTs.getMillis());
+        // If this is the first element in the batch or if the timer's output timestamp needs
+        // modifying, then set a
+        // timer.
+        if (num == 1 || minBufferedTs.read() != oldOutputTs) {
+          long targetTs =
+              MoreObjects.firstNonNull(
+                  timerTs.read(),
+                  bufferingTimer.getCurrentRelativeTime().getMillis()
```

Comment:
I might be misunderstanding something but why can't we just set this to "current time" + "maxBufferingDuration" ?

Reply:
That is what's happening - getCurrentRelativeTime is basically current time. We don't want to query the OS for current time because then unit tests won't work. Unit tests (that use TestStream) inject a fake clock so that they can control the advancement of time, and that won't work if we query the OS itself.